File: apps/android/app/src/main/java/network/thetech/fleetwright/Reassurance.kt
package network.thetech.fleetwright

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics

/**
 * The line that says nothing needs you, and why it is confident of that.
 *
 * docs/psychology.md names this as the product's real job and neither app was
 * doing it:
 *
 * > The product's real job is to convert unbounded anxiety into bounded
 * > knowledge — and the important consequence is that *"nothing needs you"* is
 * > the most important state in the system, not the least.
 *
 * A list of rows is not that. Reading five rows and concluding that none of
 * them is asking anything is work, and it is work a person redoes every time
 * they open the app — which is exactly the loop the anxiety runs in.
 *
 * SILENCE HAS TO BE TRUSTWORTHY BEFORE IT IS COMFORTABLE (§7). So this never
 * says "all good" from an absence. It counts what it can see and NAMES WHAT IT
 * CANNOT: a fleet with no health at all is a different fact from a healthy
 * fleet with nothing running, and a summary that cannot tell them apart is
 * worse than no summary.
 *
 * Matches the iOS banner clause for clause.
 */
data class Reassurance(
    val waiting: Int,
    val running: Int,
    val quiet: Int,
    val unwell: List<String>,
    val blind: Boolean,
    val healthy: Int,
) {
    companion object {
        fun of(sessions: List<Fleet.Session>, hosts: List<Fleet.FleetHost>): Reassurance {
            val unwell = hosts.filter { it.state != "healthy" }.map { it.hostId }
            return Reassurance(
                waiting = sessions.count { it.prompt != null },
                running = sessions.count { it.isRunning },
                // Only the ones that look STALLED. Counting finished sessions
                // as "quiet a while" told somebody three things needed
                // attention on a fleet where everything had gone perfectly.
                quiet = sessions.count { it.looksStalled },
                unwell = unwell,
                blind = hosts.isEmpty(),
                healthy = hosts.size - unwell.size,
            )
        }
    }

    /** The headline. One clause, and the most urgent true one. */
    val headline: String get() = when {
        waiting == 1 -> "One session is waiting for you"
        waiting > 1 -> "$waiting sessions are waiting for you"
        unwell.size == 1 -> "One machine needs a look"
        unwell.size > 1 -> "${unwell.size} machines need a look"
        blind -> "No machines are reporting"
        running == 0 -> "Nothing is running"
        else -> "Nothing needs you"
    }

    /**
     * WHY it is confident, which is the half that does the work. A headline
     * with no basis is a reassurance somebody has to take on faith, and the
     * whole argument for this line is that they should not have to.
     */
    val basis: String get() {
        if (blind) {
            return "The coordinator has no health from any machine, so this cannot say whether anything is running."
        }
        val parts = mutableListOf<String>()
        if (running > 0) parts += if (running == 1) "1 session running" else "$running sessions running"
        if (quiet > 0) parts += if (quiet == 1) "1 of them quiet a while" else "$quiet of them quiet a while"
        if (unwell.isNotEmpty()) parts += unwell.joinToString(", ")
        else parts += if (healthy == 1) "1 machine healthy" else "$healthy machines healthy"
        return parts.joinToString(" · ")
    }

    /**
     * Whether the quiet can be vouched for, which is what decides how loud the
     * card is. Calm is LOW CONTRAST on purpose: claims that hold recede, claims
     * that fail come forward. Matches `Reassurance.settled` in Swift.
     */
    val settled: Boolean get() = waiting == 0 && !blind && unwell.isEmpty()
}

/**
 * Deliberately one line and a smaller second one. This is what somebody reads
 * before deciding whether to read anything else, and a summary that takes as
 * long to read as the list it summarises has failed.
 */
@Composable
fun ReassuranceBanner(summary: Reassurance, modifier: Modifier = Modifier) {
    // NEVER COLOUR ALONE (§5). The headline always carries the meaning; this
    // only agrees with it, and a reader who cannot see it loses nothing. The
    // tones are the design system's, so amber means the same thing here, on the
    // console and on the iPhone.
    val tint: Color = when {
        summary.waiting > 0 -> Design.Palette.attention.now
        summary.blind || summary.unwell.isNotEmpty() -> Design.Palette.bad.now
        else -> Design.Palette.inkDim.now
    }
    Column(
        modifier
            .fillMaxWidth()
            // A CARD THAT IS ASKING SOMETHING WEARS ITS OWN RING. Settled, the
            // ring is the hairline every other card has; unsettled, it is the
            // tone the headline is already carrying. Calm recedes, trouble
            // comes forward, and neither depends on the colour being seen.
            .fleetCard(ring = if (summary.settled) Design.Palette.ring.now else tint)
            .padding(Design.Space.groupTight)
            // One announcement rather than two fragments: this is the line on
            // the screen worth hearing first.
            .semantics(mergeDescendants = true) {
                contentDescription = "${summary.headline}. ${summary.basis}"
            },
        verticalArrangement = Arrangement.spacedBy(Design.Space.inside),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(Design.Space.insideTight)) {
            // 26sp, and it is the only thing on the screen set that big. This
            // line is read first or the app has not done its job.
            Text(
                summary.headline,
                style = Design.Style.greeting,
                // Ink when the quiet holds, the tone when it does not. Calm
                // reads as ordinary text; trouble is the only thing on this
                // screen that is coloured for its own sake — and the sentence
                // says which it is either way.
                color = if (summary.settled) Design.Palette.ink.now else tint,
            )
        }
        Text(summary.basis, style = Design.Style.bodySmall, color = Design.Palette.inkDim.now)
    }
}

File: apps/ios/Fleetwright/Reassurance.swift
import SwiftUI

/// The line that says nothing needs you, and why it is confident of that.
///
/// docs/psychology.md names this as the product's real job and neither app was
/// doing it:
///
/// > The product's real job is to convert unbounded anxiety into bounded
/// > knowledge — and the important consequence is that *"nothing needs you"* is
/// > the most important state in the system, not the least. It is the state a
/// > person is in ninety-five percent of the time, and a surface that only
/// > becomes useful when something is wrong leaves the anxiety exactly where it
/// > was.
///
/// A list of rows is not that. Reading five rows and concluding that none of
/// them is asking anything is work, and it is work a person does again every
/// time they open the app — which is exactly the loop the anxiety runs in.
///
/// SILENCE HAS TO BE TRUSTWORTHY BEFORE IT IS COMFORTABLE (§7). So this never
/// says "all good" from an absence. It counts what it can see and NAMES WHAT IT
/// CANNOT: a host that is degraded, or that has stopped reporting, is the
/// difference between "nothing is happening" and "I have lost the ability to
/// tell you", and a summary that cannot tell them apart is worse than no
/// summary at all.
///
/// No badge, no streak, no number that moves for its own sake — see the end of
/// that document. Everything here is a fact somebody would act on.
struct Reassurance {
    let waiting: Int
    let running: Int
    let quiet: Int
    /// Hosts that are reporting something other than health, with their own
    /// words for it.
    let unwell: [String]
    /// True when we have no host report at all — which is not the same as a
    /// healthy fleet with nothing running.
    let blind: Bool
    /// Machines reporting normally.
    let healthy: Int

    init(sessions: [Fleet.Session], hosts: [Fleet.FleetHost]) {
        waiting = sessions.filter { $0.prompt != nil }.count
        running = sessions.filter { $0.isRunning }.count
        // Only the ones that look STALLED. Counting finished sessions as
        // "quiet a while" told somebody three things needed attention on a
        // fleet where everything had gone perfectly.
        quiet = sessions.filter { $0.looksStalled }.count
        unwell = hosts.filter { $0.state != "healthy" }.map { $0.hostId }
        blind = hosts.isEmpty
        healthy = hosts.count - hosts.filter { $0.state != "healthy" }.count
    }

    /// The headline. One clause, and the most urgent true one.
    var headline: String {
        if waiting > 0 { return waiting == 1 ? "One session is waiting for you" : "\(waiting) sessions are waiting for you" }
        if !unwell.isEmpty { return unwell.count == 1 ? "One machine needs a look" : "\(unwell.count) machines need a look" }
        if blind { return "No machines are reporting" }
        if running == 0 { return "Nothing is running" }
        return "Nothing needs you"
    }

    /// WHY it is confident, which is the half that does the work. A headline
    /// with no basis is a reassurance somebody has to take on faith, and the
    /// whole argument for this line is that they should not have to.
    var basis: String {
        if blind {
            return "The coordinator has no health from any machine, so this cannot say whether anything is running."
        }
        var parts: [String] = []
        if running > 0 { parts.append(running == 1 ? "1 session running" : "\(running) sessions running") }
        if quiet > 0 { parts.append(quiet == 1 ? "1 of them quiet a while" : "\(quiet) of them quiet a while") }
        if !unwell.isEmpty { parts.append(unwell.joined(separator: ", ")) }
        else { parts.append(hostsPhrase) }
        return parts.joined(separator: " · ")
    }

    private var hostsPhrase: String { healthy == 1 ? "1 machine healthy" : "\(healthy) machines healthy" }

    /// Worth a colour — and never colour alone (§5). The word above always
    /// carries the meaning; this only agrees with it.
    var tint: Color {
        if waiting > 0 { return .orange }
        if blind || !unwell.isEmpty { return .red }
        return .secondary
    }

    var symbol: String {
        if waiting > 0 { return "hand.raised" }
        if blind || !unwell.isEmpty { return "exclamationmark.triangle" }
        if running == 0 { return "moon.zzz" }
        return "checkmark.circle"
    }
}

/// The view. Deliberately one line and a second smaller one — this is the thing
/// somebody reads before deciding whether to read anything else, and a summary
/// that takes as long to read as the list it summarises has failed.
struct ReassuranceBanner: View {
    let summary: Reassurance

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 8) {
            Image(systemName: summary.symbol)
                .foregroundStyle(summary.tint)
            VStack(alignment: .leading, spacing: 2) {
                Text(summary.headline).font(.subheadline.weight(.medium))
                Text(summary.basis).font(.caption).foregroundStyle(.secondary)
            }
            Spacer(minLength: 0)
        }
        .padding(.vertical, 2)
        // One announcement rather than four fragments, because this is the one
        // line on the screen worth hearing first.
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(summary.headline). \(summary.basis)")
    }
}
